README.md
Flash Redirection
=================

Architecture
============

Objective
---------

The objective is to be able to view the flash plugin in the local desktop
for the QVD client using a streaming channel.

Documentation and repositories
------------------------------

 * nspluginwrapper repository in https://github.com/theqvd/nspluginwrapper
 * Adobe flash player
 * X2GO gathering presentation. See the doc/QVD_presentation_x2go_gathering_flash_20150621.pdf
 * Nxagent window id and event redirection patch. See https://github.com/theqvd/qvd-nx/tree/flash_plugin
 * QVD client slaveserver with nsplugin support. See ext/QVD-Client in https://github.com/theqvd/theqvd

Components
----------

 * The remote desktop. This one contains:

   * Browser. We use firefox
   * The original plugin. We have been using flash, adobe-flashplugin in Ubuntu : librery /usr/lib/adobe-flashplugin/libflashplayer.so) [This is needed to register the plugin but it is not used besides the registration with npconfig]
   * nspluginwrapper. We use https://github.com/theqvd/nspluginwrapper
 * The local desktop
   * The original plugin. We have been using flash, adobe-flashplugin in Ubuntu : librery /usr/lib/adobe-flashplugin/libflashplayer.so)
   * nspluginwrapper. We use https://github.com/theqvd/nspluginwrapper
    $HOME/.mozilla/plugins/npwrapper.libflashplayer.so. The client side is known as npw-viewer.
   * The web server. It is responsible of:
       * Start npwviewer
       * Start plugin
       * Stop plugin
 * Communication channels.
   * HTTP to invoke the plugin
   * This HTTP connection is then upgraded for socket communication (npw-wrapper in the QVD desktop and npw-viewer in the local desktop)

 ![nsplugin wrapper components](doc/componentes_flash.png)

Note:

 * The white squares represent displays
 * The grey squares represent processes
 * The blue mean libraries or modules inside the process
 * The red squares represent windows

Installation
============

In the remote desktop
---------------------

 * Install the patched agent from https://github.com/theqvd/qvd-nx/tree/flash_plugin

 * Install the flash plugin:
{{{
echo 'deb http://archive.canonical.com/ubuntu trusty partner' |sudo tee -a /etc/apt/sources.list
sudo apt-get update
sudo apt-get install adobe-flashplugin
}}}
 * Compile nspluginwrapper. '''Note:''' By default it always compiles for 32 bit to compile for 64bits it is necessary to run configure as "./configure --target-cpu=x86_64"
{{{
sudo apt-get install libglib2.0-dev libgtk2.0-dev libcurl4-openssl-dev libxt-dev
git clone https://github.com/theqvd/nspluginwrapper.git
cd nspluginwrapper
make clean && ./configure --target-cpu=x86_64 && make && sudo make install
}}}
 * Create wrapper for /usr/lib/adobe-flashplugin/libflashplayer.so
{{{
./npconfig -r /home/nito/.mozilla/plugins/npwrapper.libflashplayer.so 2>/dev/null
./npconfig -i /usr/lib/adobe-flashplugin/libflashplayer.so
sudo rm /usr/lib/mozilla/plugins/flashplugin-alternative.so
}}}
 * Verify that the plugin is activated with npconfig and firefox
{{{
$ ./npconfig -l
/home/nito/.mozilla/plugins/npwrapper.libflashplayer.so
  Original plugin: /usr/lib/adobe-flashplugin/libflashplayer.so
  Plugin viewer: /usr/lib/nspluginwrapper/i386/linux/npviewer
  Wrapper version string: 1.4.4.1-1
$
}}}
 * Execute firefox with about:plugins and verify that the plugin is activated

 ![Firefox about plugins](doc/Mozilla_Firefox_about_plugins.jpg)

 * Activate the slave channel for vma. Add the following to /etc/qvd/vma.conf. The redirection is done at port 11100 in the container, and invokes client.slave.command defined in client.conf of the QVD client.
{{{
vma.slave.enable=1
}}}

In the local desktop
--------------------


 * Install the flash plugin:
{{{
sudo apt-get install adobe-flashplugin
}}}
 * Compile nspluginwrapper. '''Note:''' By default it always compiles for 32bits, to compile for 64bits it is necessary to launch  configure as "./configure --target-cpu=x86_64"
{{{
git clone https://github.com/theqvd/nspluginwrapper.git
cd nspluginwrapper
make clean && ./configure --target-cpu=x86_64 && make && sudo make install
}}}
 * Create the wrapper for /usr/lib/adobe-flashplugin/libflashplayer.so
{{{
./npconfig -r /home/nito/.mozilla/plugins/npwrapper.libflashplayer.so 2>/dev/null
./npconfig -i /usr/lib/adobe-flashplugin/libflashplayer.so
sudo rm /usr/lib/mozilla/plugins/flashplugin-alternative.so
}}}
 * Verify that the plugin is activated with npconfig ahd with firefox
{{{
$ ./npconfig -l
/home/nito/.mozilla/plugins/npwrapper.libflashplayer.so
  Original plugin: /usr/lib/adobe-flashplugin/libflashplayer.so
  Plugin viewer: /usr/lib/nspluginwrapper/i386/linux/npviewer
  Wrapper version string: 1.4.4.1-1
$
}}}

 * Install the latest QVD-Client (version >= 0.2) and QVD-HTTPD package from github.

{{{
git clone https://github.com/theqvd/theqvd.git
cd theqvd/ext/QVD-Client
/usr/lib/qvd/bin/perl Build.PL  && ./Build && sudo ./Build install
cd ../QVD-HTTPD
/usr/lib/qvd/bin/perl Makefile.PL && make && sudo make install
}}}

 * In the $HOME/.qvd/client.conf include the line.
{{{
client.slave.command=/usr/lib/qvd/bin/qvd-client-slaveserver
}}}


Invocation secuence
===================

To test the invocation follow these steps:

 * Launch firefox in the remote Desktop. You can launch them with the following environment variables

{{{
export NPW_INIT_TIMEOUT=300
export NPW_MESSAGE_TIMEOUT=300
export NPW_DEBUG=1
export NPW_LOG=/tmp/npw-wrapper.out
firefox http://www.youtube.com/watch?v=Cp_ut3TrzxU
}}}

 * The following happens:
   * In the remote desktop the firefox detects the flash content e invokes the plugin associated plugin (nspluginwrapper).
   * The nspluginwrapper executes executes in the local desktop the following command to get the information for the plugin
{{{
/usr/lib/nspluginwrapper/i386/linux/npviewer --info --plugin /usr/lib/adobe-flashplugin/libflashplayer.so
PLUGIN_NAME 15
Shockwave Flash
PLUGIN_DESC 25
Shockwave Flash 11.2 r202
PLUGIN_MIME 98
application/x-shockwave-flash:swf:Shockwave Flash;application/futuresplash:spl:FutureSplash Player
}}}

 * In the remote desktop. The nspluginwrapper connects to localhost 11100 and invokes the following remote method, the same channel is used afterwards as socket communicationx:
{{{
GET /nsplugin?plugin=flash HTTP/1.1
Connection: Upgrade
Upgrade: qvd:slave/1.0

HTTP/1.1 101 Switching Protocols

}}}
 * In the local desktop. The process /usr/lib/qvd/bin/qvd-client-slaveserver (See TODO fix url https://github.com/theqvd/theqvd) detects the path /nsplugin and invokes npw-viewer:

 * In the local desktop. The web agent invokes npw-viewer. This knows that it is a remote invokation because of the remote-invocation flag --remote-invocation, the remote rpc port use is now deprecated (arguments --conection=127.0.0.1:2111) and is no longer used.
{{{
/usr/lib/nspluginwrapper/x86_64/linux/npviewer --plugin /usr/lib/adobe-flashplugin/libflashplayer.so --connection 127.0.0.1:2111 --remote-invocation id=libflashplayer.so-9509
}}}
 * In the remote desktop. The npw-wrapper obtains from firefox the windowid where the window should be placed (this is part from the mozilla plugins NPAPI API). From this windowid it tries to obtain the X11 Window property (xprop) _QVD_CLIENT_WID that maps the windowid in the remote desktop to the windowid in the local desktop (this can be reproduced with the command "xprop" selecting the window in the remote desktop)

 * In the remote desktop->local desktop. npw-wrapper indicates that the window needs to be created with the windowid specified by the _QVD_CLIENT_WID property (this is the local window)
 * In the remote desktop->local desktop. npw-wrapper passes all the API requests from firefox through the TCP socket, and npw-viewer access the remote URLs also through the same socket.

References
==========

* https://developer.mozilla.org/en/plugins api del firefox

* http://web.archive.org/web/20080507163354/gwenole.beauchesne.info/en/projects/nspluginwrapper nspluginwrapper

* http://library.gnome.org/devel/gdk/stable/gdk-Windows.html gdk_window_reparents


Environment variables for nspluginwrapper
=========================================

 * QVD_NSPLUGIN_DISABLE: If defined npw-wrapper will not try to use the NX channel via http://localhost:11100, insted it uses the traditional nspluginwrapper of using the unix socket. Example: "export QVD_NSPLUGIN_DISABLE=1" and to deactivate "unset QVD_NSPLUGIN_DISABLE"
 * QVD_REMOTE_INVOCATION_URL: This is to change the invocation URL. This is useful to use a different communication method instead of NX, for example SSH tunneling or direct communication. Example "export QVD_REMOTE_INVOCATION_URL=http://10.0.0.1:11200/nsplugin". To deactivate it  "unset QVD_REMOTE_INVOCATION_URL". Example to use it with socat.
{{{
socat TCP4-LISTEN:11200,fork,reuseaddr,nodelay EXEC:/usr/lib/qvd/bin/qvd-client-slaveserver
}}}
 * NPW_INIT_TIMEOUT. Connection timeout fot the TCP redirection. Normally not necessary, it used in the npw-wrapper.
 * NPW_MESSAGE_TIMEOUT. RPC response timeout for the nspluginwrapper protocol.
 * NPW_DEBUG. Set the 1 to activate the debug, this can be used in the local and remote desktop. See also NPW_LOG.
 * NPW_LOG. File where the log information is sent, if not defined it will use stderr
 * QVD_REMOTE_INVOCATION_TIMEOUT: Timeout in seconds for the first curl connection in the npw-wrapper


How to use it with SSH tunneling
================================

In the local desktop

{{{
PORT=11200
REMUSER=nito
CONTAINERIP=10.3.15.254
qvd-client &
ssh -N -R $PORT:localhost:$PORT $REMUSER@$CONTAINERIP &
socat TCP4-LISTEN:$PORT,fork,reuseaddr,nodelay EXEC:/usr/lib/qvd/bin/qvd-client-slaveserver
}}}

In the remote desktop

{{{
PORT=11200
export QVD_REMOTE_INVOCATION_URL=http://10.3.15.1:$PORT/nsplugin
export QVD_REMOTE_INVOCATION_URL=http://localhost:$PORT/nsplugin
firefox https://www.youtube.com/watch?v=4lM5nOveeA0
}}}

TODO
====

 * There seems to be some blocking in the remote invocation calls. This makes the videos stop, further investigation is needed here.
 * Test windows client
 * Warnings between GdkWindows to GTKWindows
{{{
(npviewer.bin:9533): GLib-GObject-WARNING **: invalid cast from `GdkWindow' to `GtkObject
(npviewer.bin:9533): GLib-GObject-WARNING **: invalid cast from `GdkWindow' to `GtkObject'

(npviewer.bin:9533): GLib-GObject-WARNING **: invalid cast from `GdkWindow' to `GtkWidget'

(npviewer.bin:9533): GLib-GObject-WARNING **: invalid cast from `GdkWindow' to `GtkObject'

(npviewer.bin:9533): GLib-GObject-WARNING **: invalid cast from `GdkWindow' to `GtkObject'

}}}

Window Plugin usage
===================
Or the chicken and egg problem

Options:

 * Use libflashplugin.so (the original acrobat plugin for linux). Problem the format is ELF and cygwin and similars use the COFF formand. The problem is how to invoke libflashpliugin from an windows environment. Advantages it uses gtk, X11 and similar technology

 * Use NPSWF32.dll (The original acrobat for windows). Problem it uses hWND (handle windows from USER32.DLL). This doesn't integrate natively and transparently with the X11 server (nxproxy, Xming, ...) Xming at least does not map the Xwindow id to window HWND...
   One option could be to create a GTK window in multiwindow mode or something similar, some coding into the X server would probably needed.

 * Use gnash, or a similar one, the problem is that some of the sites do not work.

 ![Firefox Window](doc/firefox_windows2.png)

 ![Firefox processes](doc/firefox_processes2.png)


One option is to try to change Xming (or whatever X server used) and allow to create X11 window that have as backend a HWND (that is a "window" in Windows world)

Some coding info
================

The first change made was to change the unix socket by a TCP socket. For that the following has been defined.

 * npw-qvd-connection.[ch]: Basic functions that save or not if you can use TCP sockets of ir you use a remote agent. If it
does not use a remote agent then fork/exec mechanism if used. If remote invocation is detected then npw-viewer is invoked via an http methd. This part is common between client and server.
   It also makes invocation http via CURL, and parses the output in format key=value. This is only executed on the client to avoid dependencies to libcurl. Here client means npw-wrapper or remote desktop.
      * translate_parent_window_id. Returns the mapping of the windowid in the nxagent, that is it retuns the property _QVD_CLIENT_WID

 * npw-wrapper.c. The function NPP_SetWindow is modified (method that is used to create and modify the windows in the plugin. Look for the functions: qvd_use_remote_plugin
 * rpc.c. The socket establishment sockets are modified. Look forqvd_use_remote_plugin

== xprop ==

Some xprop examples

To set the property, execute in the remote display (0x200006 is the container window and 0x5e00027 is the window where the events will be sent):
{{{
xprop -id 0x200006 -f _QVD_REDIRECT_EVENT 32x -set ATOM 0x5e00027
xprop -id 0x200003 -f _QVD_REDIRECT_EVENT 32x -remove _QVD_REDIRECT_EVENT
}}}

Copyright
=========
Copyright 2009-2010 by Qindel Formacion y Servicios S.L. for the QVD project
http://theqvd.com

This program is free software; you can redistribute it and/or modify it
under the terms of the GNU GPL version 3 as published by the Free
Software Foundation.

See also the original license in the README file

NEWS
nspluginwrapper NEWS -- history of user-visible changes.  2011-06-30
Copyright (C) 2005-2009 Gwenole Beauchesne
          (C) 2011 David Benjamin
          (C) 2013 Nito Martinez 

Version 1.4.4.1 - 26.Dic.2013
* Allow to use remote plugins, by using TCP sockets instead of Unix sockets.
* Need to set the following environment flags to activate it:
  * NPW_NX_AGENT_URI=http://172.26.8.82:3000/translate . This provides given
    a local window id the remote window to be used for the plugin
  * NPW_REMOTE_AGENT_URI=http://172.26.8.82:3000
    This provides some methods such as
    * /get_rpc_host_port that provides the port to be used
    * /execute_plugin to execute the remote plugin with npviewer.bin
    * /kill_plugin to kill the plugin remotely

Version 1.4.4 - 30.Jun.2011
* Fix crash in some cases when the number of watched file descriptors decreases

Version 1.4.2 - 04.Jun.2011
* Fix crash in WebKit/GTK when npwrapper.so is incorrectly treated as a plugin
* Fix symbol versioning issues with _Unwind_GetIPInfo on some systems
* Fix install process with parallel make

Version 1.4.0 - 15.May.2011
* Report capabilities over RPC to fix logic based on NULL plugin/browser hooks
* Fix initialization bug that causes Flash 10.3 to report a version of 0
* Implement ClearSiteData NPAPI extension for managing Flash LSOs
* Work around bug in Konqueror that prevents plug-ins from functioning properly

Version 1.3.2 (BETA) - 23.Apr.2011
* Work around lack of client-side windows in Flash
* Fix race condition when NPP_Destroy was called while viewer is busy
* Fix build on modern Linux platforms
* Support NPAPI 0.24, in particular, Flash can now query for private browsing
* Don't export any symbols but those necessary
* Support XEmbeding npplayer into another application
* Remove NPClass::HasMethod cache; it was incorrect
* Fix initializing two wrapped plugins with the same name in the same process
* Work around Qt bug that breaks npplayer when Kopete is installed
* Release implicit grabs before forwarding events to fix Flash context menu hang
* Work around Firefox quirk that broke windowless Flash in Firefox 4
* Bind wrapper and viewer event loops together to avoid many many race conditions
* Redesign NPRuntime bridge to avoid leaking proxy objects
* Incorportate release number into ident string so update works on distro patches

Version 1.3.0 (BETA) - 02.Jan.2009
* Don't poll for Xt events in Gtk (XEMBED) plug-ins
* Use 40 Hz timer for Xt events only when necessary (Xt input sources)
* Add NPIdentifier and NPClass::HasMethod caches, i.e. lower RPC traffic
* Add support for multiple viewer paths, see --viewer-paths=PATH-EXPR
* Add basic checks for malloc()'ed buffer underflow/overflow
* Add checks for single-threaded calls into the browser (NPN_*() functions)

Version 1.2.2 - 02.Jan.2009
* Fix support for the VLC plug-in
* Fix memory deallocation in NPN_GetStringIdentifiers()
* Fix return value if stream creation failed in standalone player

Version 1.2.0 - 26.Dec.2008
* Drop obsolete mkruntime scripts
* Use valgrind if NPW_USE_VALGRIND=yes
* Add support for SunStudio compilers
* Add support for Flash Player 10 on OpenSolaris 2008.11
* Fix build on non-Linux platforms
* Fix NPP_Destroy() to keep NPP instances longer
* Fix NPP_Destroy() to destroy the plugin window immediately

Version 1.1.10 (BETA) - 08.Dec.2008
* Fix NPPVpluginScriptableNPObject::Invalidate()
* Fix condition for delayed NPN_ReleaseObject() call
* Fix XEMBED (rework for lost events/focus regressions)
* Fix RPC for calls initiated by the plugin (SYNC mode)
* Fix invalid RPC after the plugin was NPP_Destroy()'ed

Version 1.1.8 (BETA) - 01.Dec.2008
* Delay NPN_ReleaseObject() if there is incoming RPC
* Improve plugins restart machinery (Martin Stransky)
* Close npviewer.bin sockets on exec()
* Close all open files on fork() (initial patch by Dan Walsh)
* Make `which` failures silent for soundwrappers (Stanislav Brabec)
* Allow direct execution of native plugins if NPW_DIRECT_EXEC is set

Version 1.1.6 (BETA) - 23.Nov.2008
* Enable glib memory hooks by default
* Lower priority of RPC events so that timeouts are triggered first
* Fix string_of_NPVariant() that could make some plugins crash
* Fix NPClass::Invoke(|Default)() args release
* Fix memory leak in NPN_GetStringIdentifiers()
* Fix NPN_ReleaseObject() that could dereference a deallocated NPObject
* Fix (sync) NPObject referenceCount when the object is passed to the plugin
* Fix plugin window resize in XEMBED hack mode
* Fix "javascript:" streams requests in standalone player
* Fix NPP_Write() and propagate negative lengths too (DiamondX plugin)

Version 1.1.4 (BETA) - 07.Nov.2008
* Fix memory leaks in NPRuntime bridge
* Fix XEMBED support (workaround Gtk2 and Firefox bugs)
* Fix DiamondX plugin with Konqueror4
* Fix NPP_URLNotify() (Bennet Yee)
* Fix NPAPI version that is exposed to the plugin

Version 1.1.2 (BETA) - 12.Oct.2008
* Add support for Open Solaris hosts
* Add support for ARM targets (Geraint North)
* Fix support for windowless plugins (Flash Player 10 rc)
* Fix various bugs in RPC code (crashes and concurrent messaging)
* Allow wrapping of native plugins through the -n|--native option

Version 1.1.0 (BETA) - 06.Jul.2008
* Add support for windowless plugins (Flash Player 10 beta 2)
* Add standalone plugins player (nspluginplayer)
* Restart plugins viewer on error (Martin Stransky)

Version 1.0.0 - 29.Jun.2008
* Don't wrap root plugins to system locations, keep them private
* Fix support for Acrobat Reader 8 (focus problems)
* Fix support for mozplugger (in full-debug mode)
* Fix support for WebKit
* Fix crashes with Flash Player 9 Update 3 (9.0.115)
* Fix build with Intel compiler
* Add support for IBM XLC compiler
* Improve error handling during RPC initialization (memory leaks)
* Improve error handling in NPP_WriteReady() and NPP_Write()

Version 0.9.91.5 (BETA) - 26.Aug.2007
* Fix a memory leak in NPP_Destroy()
* Fix DiamondX XEmbed example plugin

src/npw-qvd-connection.h
/*
 * npw-qvd-connection.h
 *
 * 21/8/2010 - Nito@Qindel.ES
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with this program; if not, write to the Free Software Foundation, Inc.,
 *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */
#ifndef NPW_QVD_CONNECTION_H
#define NPW_QVD_CONNECTION_H

#ifdef __cplusplus
extern "C" {
#endif
#include <X11/Xlib.h>


#define QVD_REMOTE_INVOCATION_DISABLE_ENV "QVD_NSPLUGIN_DISABLE"
#define QVD_URL_MAX_SIZE 256
#define QVD_REMOTE_INVOCATION_LOCATION "/nsplugin"
#define QVD_REMOTE_INVOCATION_BASEURL "http://localhost:11100"
#define QVD_REMOTE_INVOCATION_URL QVD_REMOTE_INVOCATION_BASEURL QVD_REMOTE_INVOCATION_LOCATION
#define QVD_REMOTE_INVOCATION_URL_ENV "QVD_REMOTE_INVOCATION_URL"
#define QVD_REMOTE_INVOCATION_TIMEOUT_ENV "QVD_REMOTE_INVOCATION_TIMEOUT"
#define QVD_XPROP_TRANSLATION "_QVD_CLIENT_WID"
#define QVD_XPROP_REDIRECT_EVENT "_QVD_REDIRECT_EVENT"
#define QVD_USERAGENT "qvd:slave/1.0"
#define QVD_CONNECT_TIMEOUT_IN_SEC 30

extern int qvd_get_client_socket();
extern int qvd_use_remote_plugin();
extern int qvd_set_remote_plugin();
extern int qvd_unset_remote_plugin();

extern int qvd_invoke_remote_plugin(const char *plugin_name);
extern int qvd_kill_plugin(int socket);
extern void *qvd_translate_wid(void *windowid);
extern void qvd_translate_xevent(XEvent *target, XEvent *source);
extern void qvd_set_xprop_qvd_redirect_event(void *windowid);

#ifdef __cplusplus
}
#endif

#endif
